src/dvc_data/fs.py

```diff
@@ -85,25 +92,28 @@ def _cache_remote_file(
         cache_storage: "ObjectStorage",
         fs: "FileSystem",
         path: "AnyFSPath",
+        hash_info: Optional["HashInfo"],
     ) -> Tuple["FileSystem", "AnyFSPath"]:
         from dvc_objects.fs.local import LocalFileSystem
 
-        if isinstance(fs, LocalFileSystem):
-            return fs, path
+        odb: "HashFileDB" = cache_storage.odb
+        oid = hash_info.value if hash_info else None
+        hash_name = hash_info.name if hash_info else None
+        assert odb.hash_name
 
-        from dvc_data.hashfile.build import _upload_file
+        if isinstance(fs, LocalFileSystem) or not oid or odb.hash_name != hash_name:
```

Comment:
@efiop, do we support external output in DataFileSystem?

Reply:
No, external outputs are not supported.